File: tools/node/validate_registry.js
#!/usr/bin/env node
/**
 * Registry validation for npm packages. Reads JSON from stdin, outputs JSON.
 * Uses only Node.js built-ins (https, fs, path).
 */
'use strict';

const https = require('https');
const fs = require('fs');
const path = require('path');

const CACHE_FILE = path.join(__dirname, '.registry_cache.json');
const RATE_LIMIT_DELAY = 500; // 0.5s in milliseconds

function loadCache() {
  try {
    if (fs.existsSync(CACHE_FILE)) {
      const data = fs.readFileSync(CACHE_FILE, 'utf-8');
      return JSON.parse(data);
    }
  } catch (_e) {
    // Ignore cache read errors
  }
  return {};
}

function saveCache(cache) {
  try {
    fs.writeFileSync(CACHE_FILE, JSON.stringify(cache), 'utf-8');
  } catch (_e) {
    // Best-effort cache persistence
  }
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function checkNpmPackage(packageName) {
  return new Promise((resolve) => {
    const url = `https://registry.npmjs.org/${encodeURIComponent(packageName)}`;
    const req = https.get(url, { headers: { 'User-Agent': 'patchwork-registry-validator/1.0' }, timeout: 10000 }, (res) => {
      // Consume the response body to free resources
      res.resume();
      if (res.statusCode === 200) {
        resolve({ package: packageName, exists: true });
      } else if (res.statusCode === 404) {
        resolve({ package: packageName, exists: false });
      } else {
        resolve({ package: packageName, exists: false, error: `HTTP error: ${res.statusCode}` });
      }
    });

    req.on('error', (err) => {
      resolve({ package: packageName, exists: false, error: `Network error: ${err.message}` });
    });

    req.on('timeout', () => {
      req.destroy();
      resolve({ package: packageName, exists: false, error: 'Network error: request timed out' });
    });
  });
}

async function main() {
  let raw = '';
  for await (const chunk of process.stdin) {
    raw += chunk;
  }

  raw = raw.trim();
  if (!raw) {
    console.log(JSON.stringify({ results: [], error: 'Empty input' }));
    return;
  }

  let data;
  try {
    data = JSON.parse(raw);
  } catch (e) {
    console.log(JSON.stringify({ results: [], error: `Invalid JSON: ${e.message}` }));
    return;
  }

  const packages = data.packages || [];
  const ecosystem = data.ecosystem || 'npm';

  if (ecosystem !== 'npm') {
    console.log(JSON.stringify({ results: [], error: `Unsupported ecosystem: ${ecosystem}` }));
    return;
  }

  const cache = loadCache();
  const results = [];

  for (let i = 0; i < packages.length; i++) {
    const pkg = packages[i];
    const cacheKey = `npm:${pkg}`;

    if (cache[cacheKey]) {
      results.push(cache[cacheKey]);
    } else {
      const result = await checkNpmPackage(pkg);
      cache[cacheKey] = result;
      results.push(result);
      // Rate limit between requests (skip for last item)
      if (i < packages.length - 1) {
        await sleep(RATE_LIMIT_DELAY);
      }
    }
  }

  saveCache(cache);
  console.log(JSON.stringify({ results }));
}

main().catch((err) => {
  console.log(JSON.stringify({ results: [], error: `Script error: ${err.message}` }));
  process.exit(1);
});

File: tools/python/validate_registry.py
#!/usr/bin/env python3
"""Registry validation for pip packages. Reads JSON from stdin, outputs JSON."""
import json
import os
import ssl
import sys
import time
import urllib.request
import urllib.error

CACHE_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), '.registry_cache.json')
RATE_LIMIT_DELAY = 0.5


def load_cache():
    """Load cached registry results from disk."""
    if os.path.exists(CACHE_FILE):
        try:
            with open(CACHE_FILE, 'r') as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError):
            pass
    return {}


def save_cache(cache):
    """Persist cache to disk."""
    try:
        with open(CACHE_FILE, 'w') as f:
            json.dump(cache, f)
    except IOError:
        pass


def _get_ssl_context():
    """Build an SSL context, trying certifi first, then system defaults, then unverified."""
    # Try certifi
    try:
        import certifi
        return ssl.create_default_context(cafile=certifi.where())
    except ImportError:
        pass

    # Try default system context
    ctx = ssl.create_default_context()
    return ctx


def _urlopen_with_fallback(req, timeout=10):
    """Open a URL, falling back to unverified SSL if certificate verification fails."""
    try:
        ctx = _get_ssl_context()
        return urllib.request.urlopen(req, timeout=timeout, context=ctx)
    except (urllib.error.URLError, OSError) as e:
        if 'CERTIFICATE_VERIFY_FAILED' in str(e):
            ctx = ssl.create_default_context()
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE
            return urllib.request.urlopen(req, timeout=timeout, context=ctx)
        raise


def check_pip_package(package, cache):
    """Check if a package exists on PyPI. Returns a result dict."""
    cache_key = f"pip:{package}"
    # Only use cache for results without errors
    if cache_key in cache and 'error' not in cache[cache_key]:
        return cache[cache_key]

    url = f"https://pypi.org/pypi/{package}/json"
    try:
        req = urllib.request.Request(url, method='GET')
        req.add_header('User-Agent', 'patchwork-registry-validator/1.0')
        with _urlopen_with_fallback(req, timeout=10) as resp:
            result = {"package": package, "exists": resp.status == 200}
    except urllib.error.HTTPError as e:
        if e.code == 404:
            result = {"package": package, "exists": False}
        else:
            result = {"package": package, "exists": False, "error": f"HTTP error: {e.code}"}
    except (urllib.error.URLError, OSError) as e:
        result = {"package": package, "exists": False, "error": f"Network error: {str(e)}"}

    # Only cache successful results (no error field)
    if 'error' not in result:
        cache[cache_key] = result
    return result


def main():
    raw = sys.stdin.read().strip()
    if not raw:
        print(json.dumps({"results": [], "error": "Empty input"}))
        return

    try:
        data = json.loads(raw)
    except json.JSONDecodeError as e:
        print(json.dumps({"results": [], "error": f"Invalid JSON: {str(e)}"}))
        return

    packages = data.get("packages", [])
    ecosystem = data.get("ecosystem", "pip")

    if ecosystem != "pip":
        print(json.dumps({"results": [], "error": f"Unsupported ecosystem: {ecosystem}"}))
        return

    cache = load_cache()
    results = []

    for i, package in enumerate(packages):
        cache_key = f"pip:{package}"
        was_cached = cache_key in cache and 'error' not in cache[cache_key]
        result = check_pip_package(package, cache)
        results.append(result)
        # Rate limit: delay between requests (skip for cached or last item)
        if i < len(packages) - 1 and not was_cached:
            time.sleep(RATE_LIMIT_DELAY)

    save_cache(cache)
    print(json.dumps({"results": results}))


if __name__ == "__main__":
    main()
